scripts/west_commands/runners/tftp.py

```diff
@@ -0,0 +1,45 @@
+# Copyright (c) 2017 Linaro Limited.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+'''Runner stub for tftp.'''
+
+import os
+from runners.core import ZephyrBinaryRunner, RunnerCaps
+
+
+class TftpBinaryRunner(ZephyrBinaryRunner):
+    '''Place-holder for tftp runner customizations.'''
+    #This is an special flash command for boards which have to copy the grub boot
+    #loader image and test case image to them, thus they can run the test case.
+    #So in this function, actually we use the network switch to control the power of
+    #board, and the board can download the necessary image from the tftp root dir
+    #via pxe boot.
+
+    def __init__(self, cfg, tftp_config):
+        super(TftpBinaryRunner, self).__init__(cfg)
+        self.tftp_config = tftp_config
+
+    @classmethod
+    def name(cls):
+        return 'tftp'
+
+    @classmethod
+    def capabilities(cls):
+        # only have to support the flash command.
+        return RunnerCaps(commands={'flash'})
+
+    @classmethod
+    def do_add_parser(cls, parser):
+        parser.add_argument('--tftp_config', help='pxe boot control command')
+
+    @classmethod
+    def create(cls, cfg, args):
+        return TftpBinaryRunner(cfg, args.tftp_config)
+
+    def do_run(self, command, **kwargs):
+        # test case image
+        strip_file = self.cfg.elf_file[:-4] + ".strip"
+        # invoke the external script set by tftp_config
+        # which will copy the iamge and complete the pxe boot
+        os.system("%s %s" %(self.tftp_config, strip_file))
```

Comment:
I don't understand what's going on here, sorry. Nothing in this file runs tftp. What's going on?

The comment above says this script "use the network switch to control the power of board, and the board can download the necessary image from the tftp root dir via pxe boot".

But this is running an arbitrary executable set up via the runner arguments in the board file, so there's nothing tftp specific here. And the runner arguments in the up_squared board file in this commit don't specify any command to run, so I don't see how this can possibly work. What am I missing?

Reply:
@mbolivar the executable set up file will be invoked via the --west-flash option in sanitycheck script, not set by up squared board file. We can use the sanitycheck --west-falsh="--tftp_config=filepath" to set the executable file path which will run the power off, copy image and power on command in it.
in this way, the tftp config file is out of the zephyr source tree, and you can realize your own pxe boot way for up board in the tftp config file, you just have to pass the file path to sanitycheck script.

Reply:
That doesn't address the concern that "nothing in this file runs tftp"

Reply:
@mbolivar In order to run all test cases automatically, we set up the up board boot by network, and set up a pxe server, if the board power on, it will download the grub boot loader image and test case image from the tftp root dir in pxe server. So the tftp there means a directory contains the test case image build in pxe server.

The argument tftp_config actually is a shell script which can control the pxe boot of up squared board remotely. In this script, there are three command,
1. power on: curl http://admin:1234@192.168.1.20/script?run003=run, we use a network switch to control the power of up board.
2. copy image: cp $0 ~/vftp/kernel, the $0 is the argument strip_file the test case image named zephyr.strip build in  pxe server, when the board power on, it will download those images automatically.
3. power on: curl http://admin:1234@192.168.1.20/script?run001=run

but maybe you can realize your own way to control the power of up board, so I just pass the file path to sanitycheck.

Reply:
I understand that you're passing the script an argument that does run tftp. However, this runner is named `tftp`, but nothing at all in the file is using tftp. It just takes an arbitrary shell command and passes it the elf file.

It's not appropriate to name this runner tftp, even if that is how you happen to be using it downstream when you run sanitycheck.

If you want to name a runner tftp and get it merged upstream, I would expect it to start up a tftp server at port 69 and make a file available to at least one client, along with some options it would take from the board file for how to reset the client and start it downloading, decide when the client is done, etc.